retrybp/filters.go

```diff
@@ -0,0 +1,132 @@
+package retrybp
+
+import (
+	"context"
+	"errors"
+	"net"
+
+	retry "github.com/avast/retry-go"
+
+	"github.com/reddit/baseplate.go/clientpool"
+)
+
+const (
+	// DefaultFilterDecision is the default decision returned by Filters at the end
+	// of the FilterChain.
+	DefaultFilterDecision = false
+)
+
+// Filters returns a `retry.RetryIf` function that checks the error against the
+// given filters and returns either the decision reached by a filter or the
+// DefaultFilterDecision.
+//
+// You should not use this with any other retry.RetryIf options as one will
+// override the other.
+func Filters(filters ...RetryFilter) retry.Option {
+	chain := &FilterChain{}
+	prev := chain
+	for i := 0; i < len(filters); i++ {
+		prev.next = &FilterChain{filter: filters[i]}
+		prev = prev.next
+	}
+	// the final chain holds a filter that returns the fallback value.
+	prev.next = &FilterChain{filter: func(_ error, _ *FilterChain) bool {
+		return DefaultFilterDecision
+	}}
+	return retry.RetryIf(chain.Next)
+}
+
+// FilterChain is an ordered chain of RetryFilters that is used by Filter to
+// determine if a function should be retried based on the error it returns.
+//
+// A FilterChain is passed to RetryFilter functions and if the filter cannot make
+// a decison about whether to retry or not, it should instead return chain.Next
+// so the next filter in the chain can check the error.
+type FilterChain struct {
+	filter RetryFilter
+	next   *FilterChain
+}
```

Comment:
nit: We usually define middleware interfaces by using two function types instead of one function type and one struct, might be better to keep the style more consistent? (the chain is essentially a middleware 😅)

Reply:
done